core/txpool/blobpool/blobpool.go

```diff
@@ -0,0 +1,1441 @@
+	}
+	// Update the metrics and return the constructed pool
+	datacapGauge.Update(int64(p.config.Datacap))
+	p.updateStorageMetrics()
+	return nil
+}
+
+// Close closes down the underlying persistent store.
+func (p *BlobPool) Close() error {
+	var errs []error
+	if err := p.limbo.Close(); err != nil {
+		errs = append(errs, err)
+	}
+	if err := p.store.Close(); err != nil {
+		errs = append(errs, err)
+	}
+	p.eventScope.Close()
+
+	switch {
+	case errs == nil:
+		return nil
+	case len(errs) == 1:
+		return errs[0]
+	default:
+		return fmt.Errorf("%v", errs)
+	}
+}
+
+// parseTransaction is a callback method on pool creation that gets called for
+// each transaction on disk to create the in-memory metadata index.
+func (p *BlobPool) parseTransaction(id uint64, size uint32, blob []byte) error {
+	item := new(blobTx)
+	if err := rlp.DecodeBytes(blob, item); err != nil {
+		// This path is impossible unless the disk data representation changes
+		// across restarts. For that ever unprobable case, recover gracefully
+		// by ignoring this data entry.
+		log.Error("Failed to decode blob pool entry", "id", id, "err", err)
+		return err
+	}
+	meta := newBlobTxMeta(id, size, item.Tx)
+
+	sender, err := p.signer.Sender(item.Tx)
+	if err != nil {
+		// This path is impossible unless the signature validity changes across
+		// restarts. For that ever unprobable case, recover gracefully by ignoring
+		// this data entry.
+		log.Error("Failed to recover blob tx sender", "id", id, "hash", item.Tx.Hash(), "err", err)
+		return err
+	}
+	if _, ok := p.index[sender]; !ok {
+		if err := p.reserve(sender, true); err != nil {
+			return err
+		}
+		p.index[sender] = []*blobTxMeta{}
+		p.spent[sender] = new(uint256.Int)
+	}
+	p.index[sender] = append(p.index[sender], meta)
+	p.spent[sender] = new(uint256.Int).Add(p.spent[sender], meta.costCap)
+
+	p.lookup[meta.hash] = meta.id
+	p.stored += uint64(meta.size)
+
+	return nil
+}
+
+// recheck verifies the pool's content for a specific account and drops anything
+// that does not fit anymore (dangling or filled nonce, overdraft).
+func (p *BlobPool) recheck(addr common.Address, inclusions map[common.Hash]uint64) {
+	// Sort the transactions belonging to the account so reinjects can be simpler
+	txs := p.index[addr]
+	if inclusions != nil && txs == nil { // during reorgs, we might find new accounts
+		return
+	}
+	sort.Slice(txs, func(i, j int) bool {
+		return txs[i].nonce < txs[j].nonce
+	})
+	// If there is a gap between the chain state and the blob pool, drop
+	// all the transactions as they are non-executable. Similarly, if the
+	// entire tx range was included, drop all.
+	var (
+		next   = p.state.GetNonce(addr)
+		gapped = txs[0].nonce > next
+		filled = txs[len(txs)-1].nonce < next
+	)
+	if gapped || filled {
+		var (
+			ids    []uint64
+			nonces []uint64
+		)
+		for i := 0; i < len(txs); i++ {
+			ids = append(ids, txs[i].id)
+			nonces = append(nonces, txs[i].nonce)
+
+			p.stored -= uint64(txs[i].size)
```

Comment:
In all other cases where we remove transactions, we also adjust the virtual balance of the sender, do we need to do this here as well?
e.g. `p.spent[addr] = new(uint256.Int).Sub(p.spent[addr], txs[i].costCap)`

Reply:
Ah nvm, we delete the spent later on anyway